crypto/crypto/src/multi_ed25519.rs

```diff
@@ -0,0 +1,547 @@
+    }
+}
+
+impl Length for MultiEd25519PublicKey {
+    fn length(&self) -> usize {
+        self.to_bytes().len()
+    }
+}
+
+impl ValidKey for MultiEd25519PublicKey {
+    fn to_bytes(&self) -> Vec<u8> {
+        self.to_bytes()
+    }
+}
+
+impl MultiEd25519Signature {
+    /// This method will also sort signatures based on index.
+    pub fn new(
+        signatures: Vec<(Ed25519Signature, u8)>,
+    ) -> std::result::Result<Self, CryptoMaterialError> {
+        let num_of_sigs = signatures.len();
+        if num_of_sigs == 0 || num_of_sigs > MAX_NUM_OF_KEYS {
+            return Err(CryptoMaterialError::WrongLengthError);
+        }
+
+        let mut sorted_signatures = signatures;
+        sorted_signatures.sort_by(|a, b| a.1.cmp(&b.1));
+
+        let mut sigvec: Vec<Ed25519Signature> = Vec::with_capacity(num_of_sigs);
+        let mut bitvec = BitVec::from_elem(MAX_NUM_OF_KEYS, false);
+
+        // Check if all indexes are unique and < MAX_NUM_OF_KEYS
+        for (i, item) in sorted_signatures.iter().enumerate() {
+            // If an index is out of range.
+            if item.1 < MAX_NUM_OF_KEYS as u8 {
+                // if an index has been set already (thus, there is a duplicate).
+                if bitvec[i] {
+                    return Err(CryptoMaterialError::BitVecError);
+                } else {
+                    sigvec[i] = item.0.clone();
+                    bitvec.set(i, true);
+                }
+            } else {
+                return Err(CryptoMaterialError::BitVecError);
+            }
+        }
+        Ok(MultiEd25519Signature(sigvec, bitvec))
+    }
+
+    /// Serialize a MultiEd25519Signature in the form of sig0||sig1||..sigN||bitvec.
+    /// Note that we omit bitvec if all bits from index = zero to index = signatures.len() - 1
+    /// are set.
+    pub fn to_bytes(&self) -> Vec<u8> {
+        let mut bytes: Vec<u8> = vec![];
+        self.0.iter().for_each(|signature| {
+            bytes.extend(signature.to_bytes().to_vec());
+        });
+        // TODO compress even more, i.e., we don't always need up to 4bytes for bitvec.
+        // No need to attach the bitvector when there is no gap between the indexes.
+        if (0..self.0.len()).any(|i| !self.1[i]) {
+            bytes.extend(self.1.to_bytes());
+        }
+        bytes
+    }
+}
+
+impl TryFrom<&[u8]> for MultiEd25519Signature {
+    type Error = CryptoMaterialError;
+
+    /// Deserialize a MultiEd25519Signature. This method will also check for bitvec validity.
+    fn try_from(bytes: &[u8]) -> std::result::Result<MultiEd25519Signature, CryptoMaterialError> {
+        let length = bytes.len();
+        // Special case for single signatures to speed up deserialization.
+        if length == ED25519_SIGNATURE_LENGTH {
+            return Ed25519Signature::try_from(bytes).map(|signature| {
+                MultiEd25519Signature(vec![signature], BITVEC_FIRST_KEY_SET.clone())
+            });
```

Comment:
That would make a fine `impl From<Ed25519Signature> for MultiEd25519Signature` you could call to here.